dev/manual_tests/lib/actions.dart

```diff
@@ -0,0 +1,442 @@
+  }
+}
+
+class SetFocusAction extends SetFocusActionBase {
+  SetFocusAction() : super('SetFocus');
+
+  @override
+  void invoke(FocusNode node, ActionTag tag) {
+    super.invoke(node, tag);
+    node.requestFocus();
+  }
+}
+
+/// Actions for manipulating focus.
+class NextFocusAction extends SetFocusActionBase {
+  NextFocusAction() : super('NextFocus');
+
+  @override
+  void invoke(FocusNode node, ActionTag tag) {
+    super.invoke(node, tag);
+    node.nextFocus();
+  }
+}
+
+class PreviousFocusAction extends SetFocusActionBase {
+  PreviousFocusAction() : super('PreviousFocus');
+
+  @override
+  void invoke(FocusNode node, ActionTag tag) {
+    super.invoke(node, tag);
+    node.previousFocus();
+  }
+}
+
+class DirectionalFocusActionTag extends ActionTag {
+  const DirectionalFocusActionTag(this.direction) : super('DirectionalFocus');
+
+  final TraversalDirection direction;
+}
+
+class DirectionalFocusAction extends SetFocusActionBase {
+  DirectionalFocusAction() : super('DirectionalFocus');
+
+  TraversalDirection direction;
+
+  @override
+  void invoke(FocusNode node, DirectionalFocusActionTag tag) {
+    super.invoke(node, tag);
+    final DirectionalFocusActionTag args = tag;
+    node.focusInDirection(args.direction);
+  }
+}
+
+/// A button class that takes focus when clicked.
+class DemoButton extends StatefulWidget {
+  const DemoButton({this.name});
+
+  final String name;
+
+  @override
+  _DemoButtonState createState() => _DemoButtonState();
+}
+
+class _DemoButtonState extends State<DemoButton> {
+  FocusNode _focusNode;
+
+  @override
+  void initState() {
+    super.initState();
+    _focusNode = FocusNode(debugLabel: widget.name);
+  }
+
+  void _handleOnPressed() {
+    print('Button ${widget.name} pressed.');
+    final UndoableActionDispatcher manager = InheritedActionDispatcher.of(context, nullOk: true);
+    setState(() {
+      manager?.invokeAction(const ActionTag('SetFocus'), node: _focusNode);
+    });
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _focusNode.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FlatButton(
+      focusNode: _focusNode,
+      focusColor: Colors.red,
+      hoverColor: Colors.blue,
+      onPressed: () => _handleOnPressed(),
+      child: Text(widget.name),
+    );
+  }
+}
+
+class FocusDemo extends StatefulWidget {
+  const FocusDemo({Key key}) : super(key: key);
+
+  @override
+  _FocusDemoState createState() => _FocusDemoState();
+}
+
+class _FocusDemoState extends State<FocusDemo> {
+  FocusNode outlineFocus;
+
+  @override
+  void initState() {
+    super.initState();
+    outlineFocus = FocusNode(debugLabel: 'Demo Focus Node');
+  }
+
+  @override
+  void dispose() {
+    outlineFocus.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final TextTheme textTheme = Theme.of(context).textTheme;
+    return InheritedShortcutMap(
+      manager: ShortcutMap(
```

Comment:
you're creating a new ShortcutMap every time you create this widget (so in principle, every build). This seems like a lot of churn. If you need something stateful, I would recommend having a stateful widget manage it, and take in an immutable description. If you don't, then just take in the immutable description and don't worry about the intermediate object.

In other words, in the build method you should just see the map (ideally a const map), not the ChangeNotifier, being built.

Reply:
OK, I've changed the `Shortcuts` widget to take an optional manager (so that it can still be overridden and replaced), and added a `shortcuts` argument that is an immutable map.